src/PersistentOrderedMap.mo

```diff
@@ -0,0 +1,966 @@
+    /// import Nat "mo:base/Nat";
+    /// import Debug "mo:base/Debug";
+    ///
+    /// let mapOps = Map.MapOps<Nat>(Nat.compare);
+    ///
+    /// let rbMap = mapOps.empty<Text>();
+    ///
+    /// Debug.print(debug_show(mapOps.size(rbMap)));
+    ///
+    /// // 0
+    /// ```
+    ///
+    /// Cost of empty map creation
+    /// Runtime: `O(1)`.
+    /// Space: `O(1)`
+    public func empty<V>() : Map<K, V> = #leaf;
+
+    /// Get an iterator for the entries of the `rbMap`, in ascending (`#fwd`) or descending (`#bwd`) order as specified by `direction`.
+    /// The iterator takes a snapshot view of the map and is not affected by concurrent modifications.
+    ///
+    /// Example:
+    /// ```motoko
+    /// import Map "mo:base/PersistentOrderedMap";
+    /// import Nat "mo:base/Nat";
+    /// import Iter "mo:base/Iter";
+    /// import Debug "mo:base/Debug";
+    ///
+    /// let mapOps = Map.MapOps<Nat>(Nat.compare);
+    /// let rbMap = mapOps.fromIter<Text>(Iter.fromArray([(0, "Zero"), (2, "Two"), (1, "One")]));
+    ///
+    /// Debug.print(debug_show(Iter.toArray(mapOps.iter(rbMap, #fwd))));
+    /// Debug.print(debug_show(Iter.toArray(mapOps.iter(rbMap, #bwd))));
+    ///
+    /// //  [(0, "Zero"), (1, "One"), (2, "Two")]
+    /// //  [(2, "Two"), (1, "One"), (0, "Zero")]
+    /// ```
+    ///
+    /// Cost of iteration over all elements:
+    /// Runtime: `O(n)`.
+    /// Space: `O(log(n))` retained memory plus garbage, see the note below.
+    /// where `n` denotes the number of key-value entries stored in the map.
+    ///
+    /// Note: Full map iteration creates `O(n)` temporary objects that will be collected as garbage.
+    public func iter<V>(rbMap : Map<K, V>, direction : Direction) : I.Iter<(K, V)>
+      = Internal.iter(rbMap, direction);
+
+    /// Returns an Iterator (`Iter`) over the key-value pairs in the map.
+    /// Iterator provides a single method `next()`, which returns
+    /// pairs in ascending order by keys, or `null` when out of pairs to iterate over.
+    ///
+    /// Example:
+    /// ```motoko
+    /// import Map "mo:base/PersistentOrderedMap";
+    /// import Nat "mo:base/Nat";
+    /// import Iter "mo:base/Iter";
+    /// import Debug "mo:base/Debug";
+    ///
+    /// let mapOps = Map.MapOps<Nat>(Nat.compare);
+    /// let rbMap = mapOps.fromIter<Text>(Iter.fromArray([(0, "Zero"), (2, "Two"), (1, "One")]));
+    ///
+    /// Debug.print(debug_show(Iter.toArray(mapOps.entries(rbMap))));
+    ///
+    ///
+    /// // [(0, "Zero"), (1, "One"), (2, "Two")]
+    /// ```
+    /// Cost of iteration over all elements:
+    /// Runtime: `O(n)`.
+    /// Space: `O(log(n))` retained memory plus garbage, see the note below.
+    /// where `n` denotes the number of key-value entries stored in the map.
+    ///
+    /// Note: Full map iteration creates `O(n)` temporary objects that will be collected as garbage.
+    public func entries<V>(m : Map<K, V>) : I.Iter<(K, V)> = iter(m, #fwd);
```

Comment:
RBTree.mo had methods `rb.entries, rb.entriesRev` and static function `RBTree.iter(rb, dir)`. Should we just consolidate on either `entries()` and `entriesRev()` or `iter(rb, dir)`?

Reply:
Related. If we keep iter(rb, dir), can we move type Direction into MapOps?

Reply:
@crusso Should we raise this question in the Slack?